### PR TITLE
python: initial implementation of ssh support

### PR DESCRIPTION
**Pull request by allisonkarlitskaya**
Add support for connecting to additional hosts via ssh, using ferny.
    
In general, this is not working well on RHEL8 — ferny depends on the KnownHostsCommand feature of OpenSSH, which is not available on RHEL8.  Without this, we don't get access to the public key of the remote host that we're connecting to, in case the host isn't known yet, which means we can't add it to the file.

- [x] #18398 

https://issues.redhat.com/browse/COCKPIT-909

### CODE REVIEW

**Comment by martinpitt** [comment]
Depending PR landed. Removing "blocked", needs rebasing now. Can't wait to see this in action!

**Comment by martinpitt** [comment]
@allisonkarlitskaya It's quite amazing to see how "small" the "remote!" commit is now. :grin: (of course, "standing on the shoulder of giants", etc., etc.) Fixing all 24 tests in one go is maybe too ambitious -- As soon as this fixes a few, let's land and make sure they don't regress again, then we can spread out investigating the others to more people (I'm happy to help).

File: src/cockpit/remote.py [hunk 1]
```diff
@@ -15,21 +15,185 @@
...
+
+    async def do_askpass(self, messages: str, prompt: str, hint: str) -> Optional[str]:
+        logger.debug('Got askpass(%s): %s', hint, prompt)
+
+        match = PasswordResponder.PASSPHRASE_RE.fullmatch(prompt)
+        if match is not None:
+            path = match.group(1)
+            logger.debug("This is a passphrase request for %s, but we don't do those.  Abort.", path)
+            self.error_message = f'locked identity: {path}'
+            return None
```
Comment (martinpitt) [inline src/cockpit/remote.py:69]: Fair enough -- Cockpit's UI does support it, but of course it's fine to implement it in a follow-up.

File: src/cockpit/remote.py [hunk 2]
```diff
@@ -15,21 +15,185 @@
...
+
+        assert self.password is not None
+        assert self.attempts == 0
+        self.attempts += 1
+        return self.password
```
Comment (martinpitt) [inline src/cockpit/remote.py:74]: So this is the case where the askpass prompt is not for an encrypted key, but the remote login PAM password, right?

File: src/cockpit/remote.py [hunk 3]
```diff
@@ -15,21 +15,185 @@
...
+        assert self.session is not None
+        logger.debug('Starting ssh session user=%s, host=%s, private=%s', self.user, self.host, self.private)
+
+        basename, colon, portstr = self.host.rpartition(':')
+        if colon and portstr.isdigit():
+            host = basename
+            port = int(portstr)
+        else:
+            host = self.host
+            port = None
```
Comment (martinpitt) [inline src/cockpit/remote.py:94]: This silently ignores cases like `self.host == 'foo:bar'`. Then `colon == ':` but the isdigit() test fails. This case should throw an error, instead of silently ignoring the `:bar` part.

File: test/pytest/test_bridge.py [hunk 4]
```diff
@@ -78,6 +78,7 @@ async def test_echo(self):
         await self.transport.assert_msg('', command='done', channel=echo)
         await self.transport.assert_msg('', command='close', channel=echo)
 
+    @unittest.skip
```
Comment (martinpitt) [inline test/pytest/test_bridge.py:81]: `@todo` perhaps? Skipping without a reason or comment feels like we're going to forget about it.

**Review by martinpitt — CHANGES_REQUESTED** [review]
Yay, I'm really happy to see some progress here, thanks! There's some things to discuss, though.

File: src/cockpit/remote.py [hunk 5]
```diff
@@ -15,21 +15,185 @@
...
-from .router import RoutingError, RoutingRule
+from ._vendor import ferny
+from .router import RoutingRule, Router
+from .peer import Peer, PeerError
+
+logger = logging.getLogger(__name__)
+
+
+class PasswordResponder(ferny.InteractionResponder):
+    PASSPHRASE_RE = re.compile(r"Enter passphrase for key '(.*)': ")
```
[the remark this answers is not in the document]
Comment (allisonkarlitskaya) [reply src/cockpit/remote.py:35]: Ya.  I'm painfully aware of this.  I tried hard to avoid it, but I don't think that we have a lot of other options here, unless we want to start patching upstream ssh....

This is the sort of stuff that got us in trouble ~5 years ago...

File: src/cockpit/remote.py [hunk 6]
```diff
@@ -15,21 +15,185 @@
...
+
+    def __init__(self, password: Optional[str]):
+        self.password = password
+        self.error_message = None
+        self.attempts = 0
+
+    async def do_hostkey(self, reason: str, host: str, algorithm: str, key: str, fingerprint: str) -> bool:
+        try:
+            logger.debug('checking ~/.ssh/known_hosts')
+            with open(os.path.expanduser('~/.ssh/known_hosts')) as file:
```
[the remark this answers is not in the document]
Comment (allisonkarlitskaya) [reply src/cockpit/remote.py:49]: This gets called, always, on every connect, sometimes multiple times, regardless of if the key is already known or not.

The manpage mentions

> Specifies a command to use to obtain a list of host keys, in addition to those listed in UserKnownHostsFile and GlobalKnownHostsFile.  This command is executed after the files have been read.

but ssh does absolutely nothing to inform us about if it found a match in those files or not, and calls the command unconditionally.

I'm glad you flagged this because I also feel a bit uncomfortable about it.  My "full strength" idea here was to check all of the files (which we can query from ssh with `-G`) and also any KnownHostsCommand that the user might have defined for themselves, but I'm not really thrilled about that either.

I think we might have to discuss the possibility of breaking compatibility with the behaviour the front-end expects from us and switch over to using an interaction that's based more on the "Do you accept this key?" prompt via askpass.  It would also let us avoid being the ones responsible for writing to the file.  More parsing, though... :/

In short, I'm not really sure what to tell you at this point other than "ya, this is gonna need to be better".

File: src/cockpit/remote.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (martinpitt) [reply src/cockpit/remote.py:35]: I think running ssh(1) through `LC_MESSAGES=C.UTF-8` would go a long way actually (but that's a ferny change, I think?)

File: src/cockpit/remote.py (hunk 1 above)
Comment (allisonkarlitskaya) [reply src/cockpit/remote.py:69]: This is actually what's supposed to happen.  We don't have support for unlocking locked keys, and we don't have any way to receive the passphrases to do so.

File: src/cockpit/remote.py (hunk 2 above)
Comment (allisonkarlitskaya) [reply src/cockpit/remote.py:74]: Yes.  And we want to make note about if we provided it or not to distinguish the report we send for the auth methods as part of the error message.

File: src/cockpit/remote.py (hunk 3 above)
Comment (allisonkarlitskaya) [reply src/cockpit/remote.py:94]: I don't agree that the port will be dropped in this case.  It will hit the `else:` branch and set `host = self.host` (ie: the full original name) and `port = None`.

Comment (martinpitt) [reply src/cockpit/remote.py:88]: Oh, ok - that feels rather strange and unpredictable, but hmm..

File: src/cockpit/remote.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (allisonkarlitskaya) [reply src/cockpit/remote.py]: It was a dumb issue: I accidentally changed the `problem='unknown-hostkey'` to `problem='unknown-host'`, which apparently means something completely different to the frontend :)

File: src/cockpit/bridge.py [hunk 7]
```diff
@@ -211,6 +242,10 @@ def main() -> None:
         print(json.dumps(Packages().get_bridge_configs(), indent=2))
         return
 
+    # The privileged bridge doesn't need ssh-agent, but the main one does
+    if 'SSH_AUTH_SOCK' not in os.environ and not args.privileged:
+        start_ssh_agent()
```
[the remark this answers is not in the document]
Comment (allisonkarlitskaya) [reply src/cockpit/bridge.py:247]: I think it's definitely impossible: I'm not sure how the environment would leak out of the one bridge and into the other...

File: src/cockpit/remote.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (martinpitt) [reply src/cockpit/remote.py]: [current run](https://cockpit-logs.us-east-1.linodeobjects.com/pull-18280-20230418-072056-1e5ebd74-fedora-37-pybridge/log.html#307) didn't finish yet, but this is clearly superior. Thanks so much!

**Comment by martinpitt** [comment]
[this run](https://cockpit-logs.us-east-1.linodeobjects.com/pull-18280-20230418-102746-de488724-centos-8-stream-pybridge/log.html#235-2) has some flakes:
 - TestServices.testBasic is issue #18465 
 - VDO failure is a too short timeout, I'll send a follow-up PR for that.

File: test/pytest/test_bridge.py (hunk 4 above)
Comment (martinpitt) [reply test/pytest/test_bridge.py:81]: I filed https://issues.redhat.com/browse/COCKPIT-983 to track this.

**Review by martinpitt — APPROVED** [review]
Good enough! Let's land this, and investigate the remaining failures. Thanks!
